src/decoder.c

```diff
@@ -0,0 +1,253 @@
+/* Copyright (C) 2023 Simo Sorce <simo@redhat.com>
+   SPDX-License-Identifier: Apache-2.0
+*/
+
+#include "provider.h"
+#include "decoder.h"
+#include "store_internals.h"
+#include "pk11_uri.h"
+#include <openssl/asn1t.h>
+#include <openssl/bio.h>
+
+typedef struct p11prov_decoder_ctx {
+    P11PROV_CTX *provctx;
+} P11PROV_DECODER_CTX;
+
+static void *p11prov_decoder_newctx(void *provctx)
+{
+    P11PROV_DECODER_CTX *dctx;
+    dctx = OPENSSL_zalloc(sizeof(P11PROV_DECODER_CTX));
+    if (!dctx) {
+        return NULL;
+    }
+
+    dctx->provctx = provctx;
+    return dctx;
+}
+
+static void p11prov_decoder_freectx(void *ctx)
+{
+    OPENSSL_clear_free(ctx, sizeof(P11PROV_DECODER_CTX));
+}
+
+static int obj_desc_verify(P11PROV_PK11_URI *obj)
+{
+    const char *desc = NULL;
+    int desc_len;
+    desc = (const char *)ASN1_STRING_get0_data(obj->desc);
+    desc_len = ASN1_STRING_length(obj->desc);
+    if (!desc || desc_len <= 0) {
+        P11PROV_debug("Failed to get description");
+        return RET_OSSL_ERR;
+    }
+
+    if (desc_len != (sizeof(P11PROV_DESCS_URI_FILE) - 1)
+        || 0 != strncmp(desc, P11PROV_DESCS_URI_FILE, desc_len)) {
+        P11PROV_debug("Description string does not match");
+        return RET_OSSL_ERR;
+    }
+    return RET_OSSL_OK;
+}
+
+static char *obj_uri_get1(P11PROV_PK11_URI *obj)
+{
+    const unsigned char *uri = ASN1_STRING_get0_data(obj->uri);
+    int uri_len = ASN1_STRING_length(obj->uri);
+    if (!uri || uri_len <= 0) {
+        P11PROV_debug("Failed to get URI");
+        return NULL;
+    }
+    char *null_terminated_uri = OPENSSL_zalloc(uri_len + 1);
```

Comment:
Use p11prov_alloc_sprintf here.

Reply:
I'm not sure if p11prov_alloc_sprintf is the right solution. The whole point of copying is to ensure it is null terminated because of the note in OSSL:

> In general it cannot be assumed that the data returned by ASN1_STRING_data() is *null terminated* or does not contain embedded nulls. 
-- https://www.openssl.org/docs/man1.1.1/man3/ASN1_STRING_get0_data.html

Reply:
p11prov_alloc_sprintf("%.*s", length, string);

Reply:
The above takes care of adding a null terminator and will not overflow